awx/main/tasks.py

```diff
@@ -429,46 +429,57 @@ def execution_node_health_check(node):
 
 
 def inspect_execution_nodes(instance_list):
-    node_lookup = {}
-    for inst in instance_list:
-        if inst.node_type == 'execution':
-            node_lookup[inst.hostname] = inst
+    with advisory_lock('inspect_execution_nodes_lock', wait=True):
+        node_lookup = {}
+        for inst in instance_list:
+            if inst.node_type == 'execution':
+                node_lookup[inst.hostname] = inst
+
+        ctl = get_receptor_ctl()
+        connections = ctl.simple_command('status')['Advertisements']
+        nowtime = now()
+        for ad in connections:
+            hostname = ad['NodeID']
+            commands = ad.get('WorkCommands') or []
+            if 'ansible-runner' not in commands:
+                continue
+            changed = False
+            if hostname in node_lookup:
+                instance = node_lookup[hostname]
+            else:
+                defaults = dict(enabled=False)
+                (changed, instance) = Instance.objects.register(hostname=hostname, node_type='execution', defaults=defaults)
+                logger.warn("Registered execution node '{}' (marked as disabled by default)".format(hostname))
+
+                default_ig = InstanceGroup.objects.get(name='default')
```

Comment:
if this doesn't exist, we would like to catch the error. An uncaught error has bad consequences.

Reply:
this should also only happened if `changed` is True.